aws-codeguruprofiler-profilinggroup/src/test/java/software/amazon/codeguruprofiler/profilinggroup/CreateHandlerTest.java

```diff
@@ -43,83 +53,226 @@
-        doThrow(ConflictException.builder().build())
-                .when(proxy).injectCredentialsAndInvokeV2(any(), any());
+            doReturn(CreateProfilingGroupResponse.builder()
+                    .profilingGroup(ProfilingGroupDescription.builder()
+                            .name(profilingGroupName)
+                            .build())
+                    .build())
+                    .when(proxy).injectCredentialsAndInvokeV2(eq(createPgRequest), any());
+        }
 
-        assertThrows(CfnAlreadyExistsException.class, () ->
-                new CreateHandler().handleRequest(proxy, request, null, logger));
-    }
+        @Test
+        public void testSuccess() {
+            final ProgressEvent<ResourceModel, CallbackContext> response = subject.handleRequest(proxy, request, null, logger);
 
-    @Test
-    public void testInternalServerException() {
-        doThrow(InternalServerException.builder().build())
-                .when(proxy).injectCredentialsAndInvokeV2(any(), any());
+            assertResponse(response);
+        }
 
-        assertThrows(CfnServiceInternalErrorException.class, () ->
-                new CreateHandler().handleRequest(proxy, request, null, logger));
+        @Test
+        public void testCorrectOperationIsCalled() {
+            subject.handleRequest(proxy, request, null, logger);
+
+            verify(proxy, times(1)).injectCredentialsAndInvokeV2(eq(createPgRequest), any());
+            verifyNoMoreInteractions(proxy);
+        }
     }
 
-    @Test
-    public void testServiceQuotaExceededException() {
-        doThrow(ServiceQuotaExceededException.builder().build())
-                .when(proxy).injectCredentialsAndInvokeV2(any(), any());
+    @Nested
+    class WithPermissions {
+
+        @BeforeEach
+        public void setup() {
+            doReturn(CreateProfilingGroupResponse.builder().build())
+                    .when(proxy).injectCredentialsAndInvokeV2(any(CreateProfilingGroupRequest.class), any());
+        }
+
+        @Nested
+        class NoPrincipals {
+
+            @Test
+            public void testNullPermissions() {
+                ResourceModel model = newResourceModel(null);
+                request = makeRequest(model);
+                assertResponse(subject.handleRequest(proxy, request, null, logger));
+            }
+
+            @Test
+            public void testNullPrincipals() {
+                ResourceModel model = newResourceModel(AgentPermissions.builder().principals(null).build());
+                request = makeRequest(model);
+                assertResponse(subject.handleRequest(proxy, request, null, logger));
+            }
+
+            @AfterEach
+            public void testCorrectOperationIsCalled() {
+                verify(proxy, times(1)).injectCredentialsAndInvokeV2(eq(createPgRequest), any());
+                verifyNoMoreInteractions(proxy);
+            }
+        }
+
+        @Nested
+        class WithPrincipals {
```

Comment:
Minor: We usually follow a pattern where we name these classes starting with "Describe" or "When". [One such example is here](https://tiny.amazon.com/jhj33g53). That case uses Kotlin, so we can afford to even use spaces, but for Java we can still write `WhenPrincipalsAreValid` or something like that.

Go wild -- test code should be as descriptive as possible :)

Reply:
Got it, renaming them to _When..._ e.g. _WhenPermissionsAreNotSet_, _WhenPrincipalsAreSet_, ...